src/hotp/mod.rs

```diff
@@ -1,152 +1,158 @@
+    generate_root(key, counter, None, None)
+}
+
+/// Layer to generate a HOTP of size N using the SHA1 hash algorithm
+///
+/// # Arguments
+///
+/// * `key` - A string of the secret
+/// * `counter` - The counter to hash
+/// * `n` - The length of the one time pad
+///
+pub fn generate_n_length(
+    key: String,
+    counter: u128,
+    n: u32,
+) -> std::result::Result<String, GenerationError> {
+    generate_root(key, counter, Some(n), None)
+}
+
+/// Layer to generate a HOTP with a custom hash digest
+///
+/// # Arguments
+///
+/// * `key` - A string of the secret
+/// * `counter` - The counter to hash
+/// * `digest` - Custom hash digest to use
+///
+pub fn generate_with_custom_digest(
+    key: String,
+    counter: u128,
+    digest: Vec<u8>,
+) -> std::result::Result<String, GenerationError> {
+    generate_root(key, counter, None, Some(digest))
+}
+
+/// Layer to generate a HOTP of size N with a custom hash digest
+///
+/// # Arguments
+///
+/// * `key` - A string of the secret
+/// * `counter` - The counter to hash
+/// * `n` - The length of the one time pad
+///
+pub fn generate_n_length_with_custom_digest(
+    key: String,
+    counter: u128,
+    n: u32,
+    digest: Vec<u8>,
+) -> std::result::Result<String, GenerationError> {
+    generate_root(key, counter, Some(n), Some(digest))
+}
+
+/// This section works to fill up the unsigned 32 bit number by:
+/// 1.  Taking the 8 bits at the offset from the digest, AND'ing with 0x7f so that we can ignore the sign bit
+/// and then bit shifting 24 to the left to fill the most significant bits.
+/// 2.  Taking the next 8 bits from the digest at (offset + 1), AND'ing with 0xff to get the set bits, shifting 16 to fill
+/// the next 8 significant bits.
+/// 3.  Same as (2.) but taking the bits from (offset + 2)
+/// 4.  Same as (2.) but taking the bits from (offset + 3)
+/// 5.  OR'ing each of these u32 so that we collapse all of the set bits into one u32
+#[doc(hidden)]
+fn generate_root(
+    key: String,
+    counter: u128,
+    digits: Option<u32>,
+    digest_arg: Option<Vec<u8>>,
+) -> std::result::Result<String, GenerationError> {
+    let defined_digits = if let Some(d) = digits { d } else { 6 };
+
+    let defined_digest = match digest_arg {
+        Some(d) => d,
+        None => match digest(key, counter, Algorithm::Sha1) {
+            Ok(d) => d,
+            _ => return Err(GenerationError::FailedToGenerateHOTP()),
+        },
+    };
+
+    let offset = if let Some(o) = defined_digest.last() {
+        o & 0xf
+    } else {
+        0
+    };
+
+    let no_offset = if let Some(o) = defined_digest.get(offset as usize) {
+        u32::from(o.clone() & 0x7f) << 24
+    } else {
+        0
+    };
+    let one_offset = if let Some(o) = defined_digest.get((offset + 1) as usize) {
+        u32::from(o.clone() & 0xff) << 16
+    } else {
+        0
+    };
+    let two_offset = if let Some(o) = defined_digest.get((offset + 2) as usize) {
```

Comment:
are there scenarios where this will be None? does the compiler let you just have the Some() unpacking? just curious

Reply:
(non blocker btw)

Reply:
Realistically, no. But the `.get(index: usize)` returns an `Option<T>` with `Some(T)` if it finds the element at the index and `None` if the vector is empty or index is out of bounds